### PR TITLE
Introduce scope stack abstraction

### LINKED ISSUE #628: Make Scope data global on desktop and mobile
This came up a few times, more recently on SO:

https://stackoverflow.com/questions/65066020/c-xamarin-forms-and-sentrysdk-context-info-not-making-it-to-sentry-errors

Currently the Hub is stored in a AsyncLocal field so data forks on new asynchronous flows.
This causes problems for Desktop and Mobile apps that want to push work to a back ground thread but would like that data available in case of an error anywhere.

This is how the Android and iOS SDKs of Sentry work, with a static global Hub.

An attempt to simulate that happened in this PR #448 but we the general approach needs to be discussed.

Resolves: #221

More context: https://forum.sentry.io/t/accessing-and-ensuring-global-shared-scope-in-net/11709/2

### LINKED ISSUE #1106: Sentry loses its scope inside the ASP.NET pipeline because it uses AsyncLocal for storage
[The official Sentry SDK docs for ASP.NET](https://docs.sentry.io/platforms/dotnet/guides/aspnet/) instructs one to start and finish tracing transactions like this in `Global.asax.cs`:

```csharp
protected void Application_BeginRequest()
{
    Context.StartSentryTransaction();
}

protected void Application_EndRequest()
{
    Context.FinishSentryTransaction();
}
```

This approach is buggy, as it will sometimes cause the Sentry scope to be lost inside the (non-Core) ASP.NET pipeline. The underlying problem is that the Sentry SDK stores its scopes in an `AsyncLocal` field in `SentryScopeManager.cs`. This is *not* a safe way to store such state in (non-Core) ASP.NET. The only safe method is likely to use `HttpContext.Items`. For more information, see:

* https://stackoverflow.com/questions/43391498/asynclocal-value-is-null-after-being-set-from-within-application-beginrequest
* https://devblogs.microsoft.com/dotnet/net-framework-4-7-1-asp-net-and-configuration-features/#asp-net-execution-step-feature
* http://piers7.blogspot.com/2005/11/threadstatic-callcontext-and_02.html

In my use case, I only care about the transaction reference, so I applied the following workaround to `Global.asax.cs`, based on the above mentioned blog post my Microsoft and how they themselves work around the issue in [Application Insights]( https://github.com/microsoft/ApplicationInsights-dotnet/blob/7c6589e2c5799edae5ffc30991b54364b2756835/WEB/Src/Web/Web/ApplicationInsightsHttpModule.cs#L150):

```csharp
public override void Init()
{
    base.Init();
    
    OnExecuteRequestStep((httpContextBase, step) =>
    {
        if (httpContextBase.CurrentNotification == RequestNotification.ExecuteRequestHandler && !httpContextBase.IsPostNotification)
        {
            SentrySdk.ConfigureScope(scope =>
            {
                if (scope.Transaction == null)
                {
                    scope.Transaction = (ITransaction) httpContextBase.Items["__SentryTransaction"];
                }
            });
        }

        step();
    });
}
```

### PR DESCRIPTION
**Pull request by Tyrrrz**
Closes #1106
Closes #628

Related to #221, #448, #505

Not entirely sure how to test this. Ideas @bruno-garcia?

### CODE REVIEW

File: test/Sentry.Tests/Internals/ScopeStack/AsyncLocalScopeStackContainerTests.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                Substitute.For<Scope>(),
+                Substitute.For<ISentryClient>()
+            );
+
+            // Act & assert
+            await Task.Run(async () =>
+            {
+                container.Stack.Should().BeNull();
+
+                await Task.Delay(10).ConfigureAwait(false);
```
Comment (bruno-garcia) [inline test/Sentry.Tests/Internals/ScopeStack/AsyncLocalScopeStackContainerTests.cs:133]: Do you need to wait for 10ms or just yield?
```suggestion
                await Task.Yield().ConfigureAwait(false);
```

**Review by bruno-garcia — COMMENTED** [review]
I like where this is going!

**Comment by bruno-garcia** [comment]
> Not entirely sure how to test this. Ideas @bruno-garcia?

I think the direction you're going for a test for each approach is a good one. Let me take another look once it's out of draft to propose some integration/unit tests

**Comment by Tyrrrz** [comment]
@bruno-garcia what do you think about the general implementation by the way? I know you said there are a lot of potential edge-cases/hidden issues, so I'm somewhat skeptical about my solution given how easy it was to implement 😅 

It could benefit to put some more responsibility inside `IScopeStackContainer`, but I decided to go with the easiest approach for now.

File: test/Sentry.Tests/Internals/ScopeStack/AsyncLocalScopeStackContainerTests.cs (hunk 1 above)
Comment (Tyrrrz) [reply test/Sentry.Tests/Internals/ScopeStack/AsyncLocalScopeStackContainerTests.cs:133]: I don't know. I thought adding a delay would introduce some potential race conditions which could be nice to test. But I guess yield is ok too.

**Review by bruno-garcia — APPROVED** [review]
Thanks!

**Comment by bruno-garcia** [comment]
We need to ship this and add some docs
